rfcs/text/0005_route_handler.md

```diff
@@ -0,0 +1,184 @@
+
+Other response parameters, such as `etag`, `MIME-type`, `bytes` that used in the Legacy platform could be adjusted via Headers.
+
+The router handler doesn't expect that logic inside can throw or return something different from `KibanaResponse`. In this case, Http service will respond with `Server error` to prevent exposure of internal logic details.
+
+#### KibanaResponseToolkit methods
+Basic primitives:
+```typescript
+type HttpResponsePayload = undefined | string | JSONValue | Buffer | Stream;
+interface HttpResponseOptions {
+  headers?: {
+   // list of  known headers
+   ...
+   // for custom headers:
+   [header: string]: number | string | string[] | undefined;
+  }
+}
+
+type StatusCode = 200 | 202 | 204 | 302 | 400 | 401 | 403 | 404 | 409 | 500;
+```
+
+##### Success
+Server indicated that request was accepted:
+```typescript
+type SuccessResponse<T> = <T extends HttpResponsePayload>(
+  payload: T,
+  options?: HttpResponseOptions
+) => KibanaResponse<T>;
+
+const kibanaResponseToolkit = {
+  ok: <T extends HttpResponsePayload>(payload: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(200, payload, options),
+  accepted: <T extends HttpResponsePayload>(payload: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(202, payload, options),
+  noContent: (options?: HttpResponseOptions) => new KibanaResponse(204, undefined, options)
+```
+
+##### Redirection
+The server wants a user to perform additional actions.
+```typescript
+const kibanaResponseToolkit = {
+  redirected: (url: string, options?: HttpResponseOptions) => new KibanaResponse(302, url, options),
+  notModified: (options?: HttpResponseOptions) => new KibanaResponse(304, undefined, options),
+```
+
+#####  Error
+Server signals that request cannot be handled and explains details of the error situation
+```typescript
+// Supports attaching additional data to send to the client
+interface ResponseError extends Error {
+  meta: {
+    data?: JSONValue;
+    code?: string; // error code to simplify search, translations in i18n, etc.
+    link?: string; // link to the docs
+  }
+}
+
+const kibanaResponseToolkit = {
+  // Client errors
+  badRequest: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(400, err, options),
+  unauthorized: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(401, err, options),
+
+  forbidden: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(403, err, options),
+  notFound: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(404, err, options),
+  conflict: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(409, err, options),
+
+  // Server errors
+  internal: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(500, err, options),
+```
+
```

Comment:
Do these above abstraction methods provide enough value to be worth their existence/DSL/maintenance? It seems like something similar to:

```ts
t.respond(options: {
  body?: HttpResponsePayload | ResponseError;
  statusCode: number;
  options?: HttpResponseOptions;
});
```
would be sufficient? Just thinking that most devs are familiar with HTTP, status codes, etc. and the DSL/abstraction may be more trouble than it's worth unless I'm missing context on why it's necessary.

Reply:
The only advantage of a DSL that I can think of is much more control over the shape of responses e.g. a 404 can't return its own custom html markup and that a 204 can't return a body. But the value in doing so might be low.

But I agree with what you said an the drawbacks in the rfc:
> `KibanaResponseToolkit` may not cover all use cases and requires an extension for specific use-cases.
> `KibanaResponseToolkit` operates low-level Http primitives, such as Headers e.g., and it is not always handy to work with them directly.

Reply:
>  The only advantage of a DSL that I can think of is much more control over the shape of responses e.g. a 404 can't return its own custom html markup and that a 204 can't return a body

right, it's easier for the platform to perform validation and we have some space to extend core functionality in future (for example, add a type definition that `WWW-Authenticate` header required when a user is rejected as unauthorized).

Reply:
I love the idea of a simple DSL for validation & consistency; for me the only thing that's missing is a clear understanding of how the DSL translates to an actual http response. (But that might just be me)

IMO that shouldn't prevent us from going this direction -- but to @jasonrhodes' point, many devs are already going to be thinking this way, so making it clear exactly what these methods are doing for them will be important.

Reply:
>  for me the only thing that's missing is a clear understanding of how the DSL translates to an actual http response payload. (But that might just be me)

For success responses, we pass payload to hapi server as is. As I understand, nothing has changed for plugin authors here.
Although, it defines different logic for the error responses. Hapi doesn't allow to respond with `Error` object, but our DSL does. For the plugin authors it looks like they throw Boom error in hapi request handler.